File: painradar/pipeline.py
"""Wire ingestion -> clustering -> scoring -> reports into one orchestration function."""
from __future__ import annotations

from typing import List, Tuple

from .clustering import cluster_posts
from .models import Post
from .opportunity import build_opportunities
from .report_html import render_html_report
from .report_json import build_report_dict
from .signals import detect_pain_signals


def run_pipeline(posts: List[Post], source_description: str) -> Tuple[dict, str]:
    pain_posts = [
        post
        for post in posts
        if detect_pain_signals(f"{post.title}\n{post.selftext}").matched_categories
    ]
    clusters = cluster_posts(pain_posts)
    opportunities = build_opportunities(clusters)
    report_dict = build_report_dict(opportunities, source_description=source_description)
    html = render_html_report(opportunities, source_description=source_description)
    return report_dict, html

File: src/pipeline.ts
import { clusterPosts } from "./clustering.js";
import { Post } from "./models.js";
import { buildOpportunities } from "./opportunity.js";
import { renderHtmlReport } from "./reportHtml.js";
import { buildReportDict } from "./reportJson.js";
import { detectPainSignals } from "./signals.js";

export interface PipelineResult {
  reportDict: Record<string, unknown>;
  html: string;
}

export function runPipeline(posts: Post[], sourceDescription: string): PipelineResult {
  const painPosts = posts.filter(
    (post) => detectPainSignals(`${post.title}\n${post.selftext}`).matchedCategories.size > 0,
  );
  const clusters = clusterPosts(painPosts);
  const opportunities = buildOpportunities(clusters);
  const reportDict = buildReportDict(opportunities, sourceDescription);
  const html = renderHtmlReport(opportunities, sourceDescription);
  return { reportDict, html };
}
